File: backend/core/domain/value_objects/hierarchy_of_controls.py
from __future__ import annotations

from backend.core.domain.value_objects.safety_enums import ControlType

# Preferred order for Hierarchy of Controls (most effective first).
HIERARCHY_OF_CONTROLS: tuple[ControlType, ...] = (
    ControlType.ELIMINATION,
    ControlType.SUBSTITUTION,
    ControlType.ENGINEERING,
    ControlType.ADMINISTRATIVE,
    ControlType.PPE,
)


def is_preferred_over(left: ControlType, right: ControlType) -> bool:
    """Return True when left is higher on the hierarchy than right."""

    return left.hierarchy_rank < right.hierarchy_rank

File: frontend/src/features/risk-assessments/utils/hierarchy-of-controls.ts
import type { ControlTypeDto } from "@/features/risk-assessments/types/risk-assessment-types";

/** Hierarchy of Controls — most preferred first (matches backend ranks). */
export const HIERARCHY_OF_CONTROLS: readonly ControlTypeDto[] = [
  "elimination",
  "substitution",
  "engineering",
  "administrative",
  "ppe",
] as const;

const RANK: Record<ControlTypeDto, number> = {
  elimination: 1,
  substitution: 2,
  engineering: 3,
  administrative: 4,
  ppe: 5,
};

export function hierarchyRank(controlType: ControlTypeDto): number {
  return RANK[controlType];
}

export function isHigherOnHierarchy(
  left: ControlTypeDto,
  right: ControlTypeDto,
): boolean {
  return hierarchyRank(left) < hierarchyRank(right);
}

export function sortControlsByHierarchy<
  T extends { controlType: ControlTypeDto },
>(controls: readonly T[]): T[] {
  return [...controls].sort(
    (a, b) => hierarchyRank(a.controlType) - hierarchyRank(b.controlType),
  );
}

export function controlTypeLabel(controlType: ControlTypeDto): string {
  const labels: Record<ControlTypeDto, string> = {
    elimination: "Устранение",
    substitution: "Замена",
    engineering: "Инженерные меры",
    administrative: "Административные меры",
    ppe: "СИЗ",
  };
  return labels[controlType];
}
